pkg/controller/vitesscluster/reconcile_topo.go

```diff
@@ -85,44 +85,56 @@ func (r *ReconcileVitessCluster) reconcileCellTopology(ctx context.Context, vt *
-	// Currently, only one region is supported so we allow routing anywhere.
-	//
-	// We also need to create the aliases before we create the cells because we
-	// don't want any vtgates to start after the cells are created but before
-	// the alias exists.
-	desiredCellsAliases := buildCellsAliases(desiredCells)
-	currentCellsAliases, err := ts.GetCellsAliases(ctx, true)
-	if err != nil {
-		r.recorder.Eventf(vt, corev1.EventTypeWarning, "TopoCellAlias",
-			"Failed to get current cell aliases: %v", err)
-		return resultBuilder.Error(err)
+	if *vt.Spec.TopologyReconciliation.RegisterCellsAliases {
+		err := r.registerCellsAliases(ctx, vt, ts, desiredCells)
+		if err != nil {
+			return resultBuilder.Error(err)
+		}
 	}
-	for alias, desiredCellsAlias := range desiredCellsAliases {
-		// If this alias already exists and matches what we are trying to update
-		// it to, skip it.
-		if _, ok := currentCellsAliases[alias]; ok {
-			if stringsets.Equal(desiredCellsAlias.Cells, currentCellsAliases[alias].Cells) {
-				continue
-			}
+
+	if *vt.Spec.TopologyReconciliation.RegisterCells {
+		err := r.registerCells(ctx, vt, ts, globalTopoImpl, desiredCells)
+		if err != nil {
+			return resultBuilder.RequeueAfter(topoRequeueDelay)
 		}
-		// Create or update cells alias.
-		err = ts.UpdateCellsAlias(ctx, alias, func(ca *topodatapb.CellsAlias) error {
-			// Even if we're creating, 'ca' will already be non-nil.
-			ca.Cells = desiredCellsAlias.Cells
-			return nil
-		})
+	}
+
+	if *vt.Spec.TopologyReconciliation.PruneCells {
+		err := r.pruneCells(ctx, vt, ts, desiredCells)
 		if err != nil {
-			r.recorder.Eventf(vt, corev1.EventTypeWarning, "TopoCellAlias",
-				"Failed to create or update cell alias: %s: %v", alias, err)
-			return resultBuilder.Error(err)
+			return resultBuilder.RequeueAfter(topoRequeueDelay)
 		}
-		r.recorder.Eventf(vt, corev1.EventTypeNormal, "TopoCellAlias",
-			"Created or updated cells alias: %s -> %v", alias,
-			desiredCellsAlias.Cells)
 	}
 
+	return resultBuilder.Result()
+}
+
+func (r *ReconcileVitessCluster) pruneCells(ctx context.Context, vt *planetscalev2.VitessCluster, ts *topo.Server, desiredCells map[string]*planetscalev2.VitessCellTemplate) error {
+	// Get list of cells in topo.
+	cellNames, err := ts.GetCellInfoNames(ctx)
+	if err != nil {
+		r.recorder.Eventf(vt, corev1.EventTypeWarning, "TopoListFailed", "failed to list cells in topology: %v", err)
+		return err
+	}
+
+	// Clean up cells that exist but shouldn't.
+	for _, name := range cellNames {
+		if desiredCells[name] == nil && vt.Status.OrphanedCells[name] == nil {
+			// The cell exists in topo, but not in the VT spec.
+			// It's also not being kept around by a blocked turn-down.
+			// See if we can delete it. This will fail if the cell is not empty.
+			if err := ts.DeleteCellInfo(ctx, name); err != nil {
+				r.recorder.Eventf(vt, corev1.EventTypeWarning, "TopoCleanupFailed", "unable to remove cell %s from topology: %v", name, err)
+				return err
```

Comment:
This used to be a "schedule requeue but don't return" so we shouldn't convert it into a return. That means this function (`pruneCells`) also needs to use its own result builder.

Reply:
Ahh good point, ok.  Will adjust.

Reply:
Adjusted - PTAL.